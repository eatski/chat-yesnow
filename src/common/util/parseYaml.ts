import { parse } from "yaml";
import { z } from "zod";

const answer = z.enum(["はい", "いいえ", "わからない"]).transform((text) => {
	switch (text) {
		case "はい":
			return "True";
		case "いいえ":
			return "False";
		case "わからない":
			return "Unknown";
	}
});

const questionExample = z.object({
	question: z.string(),
	answer,
	supplement: z.string().optional(),
	customMessage: z.string().optional(),
});

export const storyInitYaml = z.object({
	title: z.string(),
	quiz: z.string(),
	truth: z.string(),
	simpleTruth: z.string(),
	questionExamples: z.array(questionExample),
});

type ParseResult =
	| {
			error: null;
			data: z.infer<typeof storyInitYaml>;
	  }
	| {
			error: string;
	  };

export const parseYaml = (yaml: string): ParseResult => {
	try {
		parse(yaml);
		const parsed = parse(yaml);
		const typed = storyInitYaml.safeParse(parsed);
		if (!typed.success)
			return {
				error: `値に不備もしくは不足があります: ${typed.error.errors
					.flatMap((e) => e.path)
					.join(",")}`,
			};
		return {
			error: null,
			data: typed.data,
		};
	} catch (e) {
		return {
			error: "YAMLの形式が不正です。",
		};
	}
};
