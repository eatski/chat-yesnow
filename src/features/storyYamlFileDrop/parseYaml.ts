import { parse } from "yaml";
import { storyInitYaml } from "./type";
import { z } from "zod";

type ParseResult =
	| {
			error: null;
			data: z.infer<typeof storyInitYaml>;
	  }
	| {
			error: string;
	  };

export const parseYaml = (yaml: string): ParseResult => {
	try {
		parse(yaml);
		const parsed = parse(yaml);
		const typed = storyInitYaml.safeParse(parsed);
		if (!typed.success)
			return {
				error: `値に不備もしくは不足があります: ${typed.error.errors
					.flatMap((e) => e.path)
					.join(",")}`,
			};
		return {
			error: null,
			data: typed.data,
		};
	} catch (e) {
		return {
			error: "YAMLの形式が不正です。",
		};
	}
};
