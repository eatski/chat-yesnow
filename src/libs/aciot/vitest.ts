import { AxiosInstance } from "axios";
import { AciotCore, CacheMode } from "./core";
import { afterAll, beforeAll } from "vitest";
import path from "path";
import filenamify from 'filenamify';

export const applyTestHooks = (
	axios: AxiosInstance,
	config: {
		mode: CacheMode;
	},
) => {
	let aciot: AciotCore | null;
	beforeAll(({ name, file }) => {
		if (!file?.name) {
			throw new Error("suite.file.name is not defined");
		}
		const parsedPath = path.parse(file.name);
		const filename = filenamify(name, { replacement: "_" });
		aciot = new AciotCore({
			cacheBasePath: path.resolve(
				process.cwd(),
				parsedPath.dir,
				"__data__",
				parsedPath.name + parsedPath.ext,
				filename
			),
		});
		aciot.applyInterceptors(axios, config.mode);
	});
	afterAll(async ({ name, tasks }) => {
		if (tasks.every((t) => t.result?.state === "pass")) {
			console.info(name, "Clearing unused cache...");
			await aciot?.clearUnusedCache();
		}
	});
};
