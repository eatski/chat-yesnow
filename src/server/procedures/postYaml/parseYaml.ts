import { parse } from 'yaml'
import { storyInput } from "./type";

export const parseYaml = (yaml: string) => {
    const parsed = parse(yaml);
    const typed = storyInput.safeParse(parsed);
    return typed;
}