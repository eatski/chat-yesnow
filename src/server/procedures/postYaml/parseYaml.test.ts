import { readFileSync } from "fs"
import { resolve } from "path"
import {describe,test,expect} from "vitest"
import { parseYaml } from "./parseYaml" 

const sampleYaml = readFileSync(resolve(process.cwd(),"samples","sample1.yaml"), "utf-8")

describe('parseYaml', () => {
    test("should parse input yaml", ()=> {
        expect(parseYaml(sampleYaml)).toMatchSnapshot()
    })
})