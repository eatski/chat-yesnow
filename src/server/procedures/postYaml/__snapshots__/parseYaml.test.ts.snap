// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`parseYaml > should parse input yaml 1`] = `
{
  "data": {
    "questionExamples": [
      {
        "answer": "いいえ",
        "question": "太郎さんのメガネには度が入っていますか？",
        "supplement": "太郎さんのメガネは伊達メガネであり、度は入っていません。",
      },
      {
        "answer": "いいえ",
        "question": "太郎さんのメガネはブルーライトカットですか？",
        "supplement": "太郎さんのメガネにブルーライトカット機能はありません。",
      },
      {
        "answer": "はい",
        "question": "太郎さんはおしゃれのためにメガネをかけていますか？",
        "supplement": "太郎さんはオシャレ好きであり、オシャレのためにメガネをかけています。",
      },
      {
        "answer": "わからない",
        "question": "太郎さんは男ですか?",
        "supplement": "太郎さんの性別については言及されていません。",
      },
    ],
    "quiz": "太郎さんは視力がとてもいいのにメガネをかけている。なぜか？",
    "simpleTruth": "太郎さんはオシャレ好きで、オシャレのために伊達メガネをかけている。",
    "title": "太郎さんのメガネ",
    "truth": "太郎さんはオシャレ好きであり、おしゃれのために伊達メガネをかけている。",
  },
  "success": true,
}
`;
